### ISSUE TITLE
Add an inspection for undefined attributes

### ISSUE THREAD

**Issue by rocec** [asciidoctor/asciidoctor-intellij-plugin#858]
## Rationale

Add an inspection for undefined attributes. This will improve the quality of documentation by, e.g.:
- identifying typos in attributes,
- accidentally using attributes where the curly brace should be escaped, e.g. in `API call /users/{id}`.

Having an inspection allows to configure the severity level, and to ignore specific instances by annotations.

The feature complements the ability to generally treat warnings as errors (see https://github.com/asciidoctor/asciidoctor-intellij-plugin/issues/857) by handling this well-defined case explicitly. See also the related [discussion](https://asciidoctor.zulipchat.com/#narrow/stream/279642-users/topic/Warnings.20as.20errors.20in.20the.20IntelliJ.20plugin) on Zulip.

## Proposed behavior

- Use an undefined attribute in an AsciiDoc file.
- The plugin currently does not detect this.
- With the new inspection, there will be an entry in the list of problems for the corresponding file/line in the source.

**Comment by ahus1** [comment]
Pre-release 0.34.2 is now available that includes this change. In this pre-release the warning is enabled by default. Depending on how many false positives this returns, it might be disabled by default in the regular release.

I made changes in several places, I hope these changes didn't break anything else.

Please give it a try and comment here with your findings.

The pre-release of the plugin is available from [GitHub releases](https://intellij-asciidoc-plugin.ahus1.de/docs/contributors-guide/user/installing-a-preview-release.html) and the [IntelliJ AsciiDoc EAP repository](https://intellij-asciidoc-plugin.ahus1.de/docs/contributors-guide/user/add-eap-repository-to-ide.html).

**Comment by rocec** [comment]
I am getting a lot of false positives in an Antora project. My attributes are defined in a partial in the ROOT module that all pages include. If I understand correctly, the plugin will currently treat all attributes as defined that are defined in any AsciiDoc page, but not those defined in a partial. IMHO, this should be changed to include definitions in partials.

Taking one step back: What is the reason to consider an attribute as defined for one page just because it is defined in another page?

**Comment by ahus1** [comment]
Thanks for taking the time to test it. The rationale behind that is to avoid false positives: when looking at a sub-page, a parent page might have set the attributes, and it might not be clear in all cases when looking at a child page how it is included in a parent page.

I need to double check, but I might have restricted this to attributes declared and used within a module. I found that otherwise auto-completion of attribute names got out of hand, as it showed attributes relevant only to other modules in the current module.

I'm not sure if I fully understand you setup: do you have a  ROOT module with a partial, and other content modules where this partial is included? To define attributes for all pages, I would have expected these attributes to be part of a playbook. 

Could you show me a (minimal) example project? Thanks!

**Comment by rocec** [comment]
Well, the thanks go to you, of course.

I have to withdraw my previous report. I failed reproducing the false positives in a new minimal project. Then I revisited the bigger project and noticed that it was "messed up" in IntelliJ. The files in src/asciidoc were not considered "project files" anymore. (I do not know how that happened.) I closed IntelliJ, deleted the .idea folder, reopened the project, et voilà: All defined attributes are now resolved, also from partials. Undefined attributes are reported as desired.

With respect to my setup, you understood correctly. I use a partial in the ROOT module to define component-wide shared attributes and settings. You are right, the playbook may be a better place to do this. My rationale is that I compile the same AsciiDoc sources not only with Antora, but also with Asciidoctor. Having commons in a partial works with both tools. There may be a cleaner way, but I guess that's another discussion and off topic here.

Summing up, the implementation looks fine so far. I will continue using it and report any additional findings here.

**Comment by mojavelinux** [comment]
> I use a partial in the ROOT module to define component-wide shared attributes and settings. You are right, the playbook may be a better place to do this. 

Also consider that you can set attributes in antora.yml. That's usually better than putting them in the playbook because they stay with the content. It's the closest thing to a shared attributes file.

**Comment by rocec** [comment]
Thanks for the tip. I have to admit that I just became aware of the possibility to set attributes in the component descriptor when I read the explanation of the new inspection earlier today.
